### DISCUSSION TITLE
AVX2 broadcast from register?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/runtime#84518] [Q&A] [answered]
If I understand correctly, all broadcast intrinsics go in the `Avx2` class. There are a bunch of overloads, but for parameters smaller than or equal to 64-bit, there are only overloads that accept memory pointers. The x86 `vpbroadcastx` (where x = b/w/d/q) instruction supports both memory form and register form. See, for example, https://www.felixcloutier.com/x86/vpbroadcast . The `Avx2` class is not providing these methods now.

There are workarounds of using the memory form, taking the address of the local variable, or using the `Create` method, but in either case, .NET cannot optimize it into the desired `vpbroadcastq` ([sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbhr0mrTj36CRGMRKm4WADQAcSOdXkBmJqQYBhBgG8aDR03Pdc2AGYwmKBgFVcMFaw2BgwABQAJhAcwAA2XhFoDFEx8QBUDAAOpACUDk721E7FDABu2FAMGPiZDAC8yQwZpMYlTuWVpfUMAGowYBjQpACsSCxBMCHh1ZlJM3M1C5k5rW0MAIKlCCwAyoOwYdlJpSv5jgC+Z84Mrh5exD7+MACyMPjQAJ4AYtD4kdFxBJJFKAjLZPJFApXYodKo1boRJoMFrQ9oVMrdTYIUgsABCUAg2AiYDwGB2JNiFQAKhA+gMhqMwgAyGanSFtLG7fbhI5lNnFS7Uc5AA)).

So what should we do here if we want to broadcast from a `double` (or a `Vector64<double>`) to a `Vector256<double>`?

**Comment by acaly** [comment]
I found an answer myself. It seems that we can use `Avx2.BroadcastScalarToVector256(Vector128.CreateScalarUnsafe(tmp))`.

But I am not sure this is good, because we must use the "Unsafe" form in order for the optimization to occur.

**Reply by EgorBo** [reply]
Unsafe here means that it is not guaranteed that all components of Vector128 except the first one will be 0, there can be some garbage values. Which is perfectly fine for your case since you're broadcasting the first component only

**Comment by tfenise** [comment, marked as answer]
There is an overload of `Vector256.Create` that takes only one parameter, which should be optimized properly.
